### ISSUE TITLE
Editor: add syntax highlighting for dialogic-script

### ISSUE THREAD

**Issue by dhowe** [dhowe/dialogic#88]
This is supported by code-mirror (see  #83) -- lets discuss once that is ready...

**Comment by cqx931** [comment]
Code Mirror is done. Are we going to use an existing language mode from Code-Mirror or do we need to create our own js file implementing the mode?
https://codemirror.net/mode/

**Comment by dhowe** [comment]
we need to create our own mode for dialogic

**Comment by cqx931** [comment]
Here is the first attempt of adding syntax highlighting:
https://github.com/dhowe/dialogic/tree/syntax

Preview:
![image](https://user-images.githubusercontent.com/3056154/42295016-7c135e8a-7fe6-11e8-9db1-397ad8f56521.png)

Most of the colors are just using the default theme or my random pick, we could talk about that later...
I was looking at the [Command Reference](https://github.com/dhowe/dialogic/wiki/Command-Reference) when I was working on this. And I'm not sure about the best way to highlight METADATA or LABEL. Any reference/specs on syntax highlighting details?

**Comment by dhowe** [comment]
The most important thing is that dialog which will be spoken is a different color than everything else. This is includes the text following OPT, ASK, and SAY commands, but does not include metadata, which should be its own color (everything between the { ... } ). Labels (including Chat names) and variables should also have a color. See below...

**Comment by dhowe** [comment]
Alpha 
- [ ] Metadata (unique color or italic?)
- [ ] Labels
- [ ] Groups 

__Beta__
- [ ] Actors

**Comment by cqx931** [comment]
Please check: https://github.com/dhowe/dialogic/pull/141
There is one case that I can't figure out how to parse within the current simpleMode framework: label following `CHAT`.

Preview:
![image](https://user-images.githubusercontent.com/3056154/43663165-6ae045da-9736-11e8-9fb1-d2ba5ddebeea.png)

**Comment by dhowe** [comment]
Also, it is unclear to me whether a group should override the highlighting of a variable inside? See '// group with variable' above...

**Comment by cqx931** [comment]
Please check: https://github.com/dhowe/dialogic/pull/141/commits/fc4060eb0506611b3b3340fe5506c614398f57c4

- Fixed variable inside a group state
- Use BOLD for commands
- After rechecking the documentation for defineeSimpleMode() in codeMirror, I implemented the label highlighting after CHAT using a new state instead of a rule. I had problem with this particular case because look behind is not supported in javascript regex.

![image](https://user-images.githubusercontent.com/3056154/43797875-d635f08c-9a56-11e8-9c45-8342e80b9f67.png)

**Comment by dhowe** [comment]
We also have the issue of variables in metadata:
`FIND {emotion=$currentEmotion}`

I wonder if we should let the larger 'thing' override the smaller, so that variables in groups are colored as the group, variables in meta are colored as the meta, etc.  ?
thoughts?

**Comment by cqx931** [comment]
Visually I think it is more clear to me if the larger type override the smaller. 
If the color coding is too scattered(especially for meta data), it won't help the user to parse the information but makes it harder.

**Comment by dhowe** [comment]
> Visually I think it is more clear to me if the larger type override the smaller.
Ok, agreed for now...

I think we need to adjust the colors a bit: groups should be close to the normal text color, as these will also appear to the user (perhaps just swap metadata and group colors?)

**Comment by cqx931** [comment]
Please check the new PR for all the features in beta
https://github.com/dhowe/dialogic/pull/149/commits

![image](https://user-images.githubusercontent.com/3056154/44475972-76528680-a604-11e8-82f2-76cf6be29ed3.png)
